File: src/com/ampro/util/Timer.java
/**
 *
 */
package com.ampro.util;

import java.util.List;

/**
 *
 * A Timer object contains a start-time (long) that is defined upon creation.<br>
 * The method getTime() returns the current duration of the timer's run time in seconds.
 *
 * @author Götten Töter
 *
 */
public class Timer {

	private long startTime = System.nanoTime();

	/**
	 * Constructing a Timer object starts
	 */
	public Timer(){}

	/**
	 * @return The current duration of the timer's run time in HH:MMM:SS
	 */
	public String formattedTime(){
		double seconds = this.getElapsedTime() / Math.pow(10, 9);
		int hours=0;
		int min=0;
		if(seconds > 60)
			for(int i=0; i < seconds; i++)
				if(i==60){
					min++;
					seconds -= 60;
					i=0;
					if(min==60){
						hours++;
						min -= 60;
					}
				}
		return "" + hours + ":" + min + ":" + seconds;
	}

	/**
	 * Resets the starting time and returns the previous duration
	 * @return
	 */
	public String reset(){
		String lastTime = this.formattedTime();
		this.startTime = System.nanoTime();
		return lastTime;
	}

	/**
	 *
	 * @param List<\Timers> or Timer[]
	 * @return
	 */
	public static String average(Object obj){

		long AverageNanoTime = 0;
		if(obj instanceof Timer[]){
			for(int i=0; i < ((Timer[]) obj).length; i++)
				AverageNanoTime += ((Timer[]) obj)[i].getElapsedTime();
			AverageNanoTime /= ((Timer[]) obj).length;
		}
		else if(obj instanceof List<?>){
			if( ((List<?>)obj).get(0) instanceof Timer){
				for(int i=0; i < ((List<?>) obj).size(); i++)
					AverageNanoTime += ((Timer) ((List<?>) obj).get(i)).getElapsedTime() ;
				AverageNanoTime /= ((List<?>) obj).size();
			}else if(((List<?>)obj).get(0) instanceof Long){
				for(int i=0; i < ((List<?>) obj).size(); i++)
					AverageNanoTime += (Long) ((List<?>) obj).get(i);
				AverageNanoTime /= ((List<?>) obj).size();
			}
		}
		else if(obj instanceof long[]){
			for(int i=0; i < ((long[]) obj).length; i++)
				AverageNanoTime += ((long[]) obj)[i];
			AverageNanoTime /= ((long[]) obj).length;
		}

		double seconds = AverageNanoTime/ Math.pow(10, 9);
		int hours=0;
		int min=0;
		if(seconds > 60)
			for(int i=0; i < seconds; i++)
				if(i==60){
					min++;
					seconds -= 60;
					i=0;
					if(min==60){
						hours++;
						min -= 60;
					}
				}
		return "" + hours + ":" + min + ":" + seconds;
	}

	public long getElapsedTime(){
		return System.nanoTime() - this.startTime;
	}

	@Override
	public String toString(){
		return this.formattedTime();
	}

}

File: src/main/kotlin/com/ampro/evemu/util/Timer.kt
package com.ampro.evemu.util

/**
 * Timer and Time-related functions, classes, and variables.
 *
 * @author Jonathan Augustine
 * @since 3.0
 */

import java.time.LocalDateTime
import java.time.format.DateTimeFormatter


/** @return The current local date and time. dd-MM-yyyy HH:mm:ss */
val NOW: String get() = LocalDateTime.now().format(
        DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm:ss"))

/** @return The current local date and time. dd-MM-yyyy HH-mm-ss */
val NOW_FILE: String get() = NOW.replace(":", "-")

/**
 * A Timer object contains a start-time (millisec) that is defined upon creation.
 *
 * @author Jonathan Augustine
 * @since 2.0
 */
class Timer(private var startTime: Long = System.currentTimeMillis()) {

    /** If the timer is running or not */
    private var running: Boolean = true

    /** The time elapsed since the Timer started (-1 the timer is running) */
    private var elapsedTime: Long = -1

    /** @return The time since the Timer started or was stopped */
    fun getElapsedTime() : Long {
        return if (running) {
            System.currentTimeMillis() - startTime
        } else {
            elapsedTime
        }
    }

    /** @return The current duration of the timer's run time in HH:MMM:SS */
    fun formattedTime(): String = format(this.getElapsedTime())

    companion object {
        fun format(millis: Long) : String {
            var seconds = (millis / 1_000).toDouble()//Math.pow(10.0, 9.0)
            var hours = 0
            var min = 0
            if (seconds > 60) {
                var i = 0
                while (i < seconds) {
                    if (i == 60) {
                        min++
                        seconds -= 60.0
                        i = 0
                        if (min == 60) {
                            hours++
                            min -= 60
                        }
                    }
                    i++
                }
            }
            return "$hours:$min:$seconds"
        }
    }

    /** @return The current duration of the timer's run time in HH:MMM:SS */
    override fun toString(): String = this.formattedTime()

    /** Start the Timer if it is not running */
    fun start() {
        this.running = true
        this.elapsedTime -1L
    }

    /**
     * Stop the timer (set the end time)
     *
     * @return The formatted duration
     */
    fun stop() : String {
        this.elapsedTime = System.currentTimeMillis()
        this.running = false
        return this.formattedTime()
    }

    /**
     * Resets the starting time and returns the previous duration
     *
     * @return the last duration formatted
     */
    fun reset() : String {
        val lastTime = this.formattedTime()
        this.startTime = System.currentTimeMillis()
        return lastTime
    }

}

/**
 * @param timers Array of Timers
 * @return The average duration of the provided Timers, formatted HH:MMM:SS
 */
fun timerAvg(timers: List<Timer>): String {

    var averageNanoTime: Long = 0

    timers.forEach { averageNanoTime += it.getElapsedTime() }

    averageNanoTime /= timers.size.toLong()

    var seconds = averageNanoTime / Math.pow(10.0, 9.0)
    var hours = 0
    var min = 0
    if (seconds > 60) {
        var i = 0
        while (i < seconds) {
            if (i == 60) {
                min++
                seconds -= 60.0
                i = 0
                if (min == 60) {
                    hours++
                    min -= 60
                }
            }
            i++
        }
    }
    return "$hours:$min:$seconds"
}
